lib/karafka/base_worker.rb

```diff
@@ -10,17 +10,18 @@ class BaseWorker
     # @param params_batch [Array] Array with messages batch
     def perform(topic_id, params_batch)
       Karafka.monitor.notice(self.class, params_batch)
-      controller(topic_id, params_batch).call
+      controller(topic_id, params_batch).perform
     end
 
     private
 
     # @return [Karafka::Controller] descendant of Karafka::BaseController that matches the topic
     #   with params_batch assigned already (controller is ready to use)
     def controller(topic_id, params_batch)
-      @controller ||= Karafka::Routing::Router.build(topic_id).tap do |ctrl|
-        ctrl.params_batch = ctrl.topic.interchanger.parse(params_batch)
-      end
+      topic = Karafka::Routing::Router.find(topic_id)
+      controller = topic.controller.new
+      controller.params_batch = topic.interchanger.parse(params_batch)
+      controller
```

Comment:
You don't have to make this change, but I always cannot resist some playful code golf :joy::

```ruby
topic.controller.new.tap { |c| c.params_batch = topic.interchanger.parse(params_batch) }
```